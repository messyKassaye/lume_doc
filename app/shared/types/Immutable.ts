import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
