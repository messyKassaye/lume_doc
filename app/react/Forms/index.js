import Select from './components/Select';
import DatePicker from './components/DatePicker';
import Captcha from './components/Captcha';
import DateRange from './components/DateRange';
import FormGroup from './components/FormGroup';
import MultiSelect, { MultiSelectTristate } from './components/MultiSelect';
import { LookupMultiSelect } from './components/LookupMultiSelect';
import MarkDown from './components/MarkDown';
import Nested from './components/Nested';
import NestedMultiselect from './components/NestedMultiselect';
import MultiDate from './components/MultiDate';
import MultiDateRange from './components/MultiDateRange';
import DropdownList from './components/DropdownList';
import IconSelector from './components/IconSelector';
import Numeric from './components/Numeric';
import NumericRange from './components/NumericRange';
import NumericRangeSlide from './components/NumericRangeSlide';
import RadioButtons from './components/RadioButtons';
import Switcher from './components/Switcher';
import Geolocation from './components/Geolocation';
import LinkField from './components/LinkField';
import FormValue from './components/FormValue';
import MediaField from './components/MediaField';
export * from './components/MultiSuggest';

export {
  Select,
  FormGroup,
  DatePicker,
  Captcha,
  DateRange,
  MultiSelect,
  MultiSelectTristate,
  MarkDown,
  Nested,
  NestedMultiselect,
  Numeric,
  NumericRange,
  NumericRangeSlide,
  MultiDate,
  MultiDateRange,
  DropdownList,
  IconSelector,
  RadioButtons,
  Switcher,
  Geolocation,
  LinkField,
  FormValue,
  LookupMultiSelect,
  MediaField,
};
