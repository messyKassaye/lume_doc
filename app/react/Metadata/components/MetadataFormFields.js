import { FormGroup } from 'app/Forms';
import t from 'app/I18N/t';
import { preloadOptionsLimit } from 'shared/config';
import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field } from 'react-redux-form';
import { propertyTypes } from 'shared/propertyTypes';
import { getSuggestions } from 'app/Metadata/actions/actions';
import { Translate } from 'app/I18N';
import {
  DatePicker,
  DateRange,
  Geolocation,
  LinkField,
  MarkDown,
  MultiDate,
  MultiDateRange,
  MultiSelect,
  MultiSuggest,
  Nested,
  Numeric,
  Select,
  LookupMultiSelect,
  MediaField,
} from '../../ReactReduxForms';
import MultipleEditionFieldWarning from './MultipleEditionFieldWarning';
import { MediaModalType } from './MediaModal';

export const translateOptions = thesauri =>
  thesauri
    .get('values')
    .map(optionIm => {
      const option = optionIm.toJS();
      option.label = t(thesauri.get('_id'), option.label, null, false);
      if (option.values) {
        option.options = option.values.map(val => ({
          ...val,
          label: t(thesauri.get('_id'), val.label, null, false),
        }));
      }
      return option;
    })
    .toJS();

export class MetadataFormFields extends Component {
  getField(property, _model, thesauris) {
    let thesauri;
    let totalPossibleOptions = 0;
    const { dateFormat, version, entityThesauris, attachments } = this.props;
    const propertyType = property.type;

    const plainAttachments = attachments.toJS();

    switch (propertyType) {
      case 'select':
        thesauri = thesauris.find(opt => opt.get('_id').toString() === property.content.toString());
        return (
          <Select
            model={_model}
            optionsValue="id"
            options={translateOptions(thesauri)}
            placeholder="Select..."
          />
        );
      case 'multiselect':
        thesauri = thesauris.find(opt => opt.get('_id').toString() === property.content.toString());
        return (
          <MultiSelect
            model={_model}
            optionsValue="id"
            options={translateOptions(thesauri)}
            prefix={_model}
            forceHoist={version === 'OneUp'}
            placeholder={
              version === 'OneUp'
                ? `${t('System', 'Search', null, false)} '${thesauri.get('name')}'`
                : null
            }
          />
        );
      case 'relationship':
        if (property.content) {
          const source = thesauris.find(
            opt => opt.get('_id').toString() === property.content.toString()
          );

          totalPossibleOptions = source.get('optionsCount');
          thesauri = translateOptions(source);
        }

        if (!property.content) {
          thesauri = Array.prototype
            .concat(
              ...thesauris
                .filter(filterThesauri => filterThesauri.get('type') === 'template')
                .map(source => {
                  totalPossibleOptions += source.get('optionsCount');
                  return translateOptions(source);
                })
            )
            .slice(0, preloadOptionsLimit);
        }

        if (entityThesauris.get(property.name)) {
          entityThesauris
            .get(property.name)
            .toJS()
            .forEach(o => {
              thesauri.push({ id: o.value, label: o.label });
            });
        }

        return (
          <LookupMultiSelect
            lookup={getSuggestions.bind(null, property.content ? [property.content] : [])}
            model={_model}
            optionsValue="id"
            options={thesauri}
            totalPossibleOptions={totalPossibleOptions}
            prefix={_model}
            sort
          />
        );
      case 'date':
        return <DatePicker model={_model} format={dateFormat} />;
      case 'daterange':
        return <DateRange model={_model} format={dateFormat} />;
      case 'numeric':
        return <Numeric model={_model} />;
      case 'markdown':
        return <MarkDown model={_model} />;
      case 'nested':
        return <Nested model={_model} />;
      case 'multidate':
        return <MultiDate model={_model} format={dateFormat} />;
      case 'multidaterange':
        return <MultiDateRange model={_model} format={dateFormat} />;
      case 'geolocation':
        return <Geolocation model={_model} />;
      case 'link':
        return <LinkField model={_model} />;
      case 'media':
        return (
          <MediaField model={_model} attachments={plainAttachments} type={MediaModalType.Media} />
        );
      case 'image':
        return (
          <MediaField model={_model} attachments={plainAttachments} type={MediaModalType.Image} />
        );
      case 'preview':
        return (
          <div>
            <em>This content is automatically generated</em>
          </div>
        );
      case 'text':
        return (
          <Field model={_model}>
            <input type="text" className="form-control" />
          </Field>
        );
      default:
        return false;
    }
  }

  render() {
    const { thesauris, template, multipleEdition, model, showSubset } = this.props;

    const mlThesauri = thesauris
      .filter(thes => !!thes.get('enable_classification'))
      .map(thes => thes.get('_id'))
      .toJS();
    const fields = template.get('properties').toJS();
    const templateID = template.get('_id');

    return (
      <div>
        {fields
          .filter(p => !showSubset || showSubset.includes(p.name))
          .map(property => (
            <FormGroup key={property.name} model={`.metadata.${property.name}`}>
              <ul
                className={`search__filter is-active ${
                  this.props.highlightedProps.includes(property.name) ? 'highlight' : ''
                }`}
              >
                <li className="title">
                  <label>
                    <MultipleEditionFieldWarning
                      multipleEdition={multipleEdition}
                      model={model}
                      field={`metadata.${property.name}`}
                    />
                    {templateID ? (
                      <Translate context={templateID}>{property.label}</Translate>
                    ) : (
                      property.label
                    )}
                    {property.required ? <span className="required">*</span> : ''}
                  </label>
                </li>
                {mlThesauri.includes(property.content) &&
                [propertyTypes.multiselect, propertyTypes.select].includes(property.type) ? (
                  <li className="wide">
                    <MultiSuggest
                      model={`.suggestedMetadata.${property.name}`}
                      selectModel={`.metadata.${property.name}`}
                      propertyType={property.type}
                    />
                  </li>
                ) : null}
                <li className="wide">
                  {this.getField(property, `.metadata.${property.name}`, thesauris)}
                </li>
              </ul>
            </FormGroup>
          ))}
      </div>
    );
  }
}

MetadataFormFields.defaultProps = {
  multipleEdition: false,
  dateFormat: null,
  version: undefined,
  showSubset: undefined,
  entityThesauris: Immutable.fromJS({}),
  attachments: Immutable.fromJS([]),
  highlightedProps: [],
};

MetadataFormFields.propTypes = {
  template: PropTypes.instanceOf(Immutable.Map).isRequired,
  model: PropTypes.string.isRequired,
  thesauris: PropTypes.instanceOf(Immutable.List).isRequired,
  multipleEdition: PropTypes.bool,
  dateFormat: PropTypes.string,
  showSubset: PropTypes.arrayOf(PropTypes.string),
  version: PropTypes.string,
  entityThesauris: PropTypes.instanceOf(Immutable.Map),
  highlightedProps: PropTypes.arrayOf(PropTypes.string),
  attachments: PropTypes.instanceOf(Immutable.List),
};

export const mapStateToProps = (state, ownProps) => {
  const { storeKey } = ownProps;

  let attachments = Immutable.fromJS([]);

  if (storeKey === 'library' || storeKey === 'uploads') {
    const selectedDocuments = state[storeKey].ui.get('selectedDocuments');
    attachments = selectedDocuments.size ? selectedDocuments.get(0).get('attachments') : undefined;
  }

  if (storeKey === 'documentView') {
    const entity = state.documentView.doc;
    attachments = entity.get('attachments');
  }

  if (!storeKey) {
    const entity = state.entityView.entity;
    attachments = entity.get('attachments');
  }

  return {
    dateFormat: state.settings.collection.get('dateFormat'),
    entityThesauris: state.entityThesauris,
    attachments,
  };
};

export default connect(mapStateToProps)(MetadataFormFields);
