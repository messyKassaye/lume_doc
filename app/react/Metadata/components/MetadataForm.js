import entitiesUtil from 'app/Entities/utils/filterBaseProperties';
import { Select as SimpleSelect } from 'app/Forms';
import { I18NLink, t, Translate } from 'app/I18N';
import { notificationActions } from 'app/Notifications';
import { FormGroup } from 'app/ReactReduxForms';
import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field, Form } from 'react-redux-form';
import { bindActionCreators } from 'redux';
import { createSelector } from 'reselect';
import { Icon } from 'UI';
import defaultTemplate from '../helpers/defaultTemplate';
import validator from '../helpers/validator';
import { wrapEntityMetadata } from '../helpers/wrapper';
import { IconField } from './IconField';
import MetadataFormFields from './MetadataFormFields';

const immutableDefaultTemplate = Immutable.fromJS(defaultTemplate);

const selectTemplateOptions = createSelector(
  s => s.templates,
  templates => templates.map(tmpl => ({ label: tmpl.get('name'), value: tmpl.get('_id') }))
);

export class MetadataForm extends Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
    this.onSubmitFailed = this.onSubmitFailed.bind(this);
  }

  componentWillUnmount() {
    this.props.componentWillUnmount();
  }

  onSubmit(entity) {
    this.props.onSubmit(
      wrapEntityMetadata(entitiesUtil.filterBaseProperties(entity)),
      this.props.model
    );
  }

  onSubmitFailed() {
    this.props.notify(t('System', 'Invalid form', null, false), 'danger');
  }

  renderTemplateSelect(templateOptions, template) {
    if (templateOptions.size) {
      return (
        <FormGroup>
          <ul className="search__filter">
            <li>
              <label>{t('System', 'Type')}</label>
            </li>
            <li className="wide">
              {this.props.initialTemplateId !== undefined &&
                this.props.initialTemplateId !== this.props.templateId && (
                  <span>
                    <Icon icon="exclamation-triangle" />
                    <Translate context="System">
                      Changing the type will erase all connections to this entity.
                    </Translate>
                  </span>
                )}
            </li>
            <li className="wide">
              <SimpleSelect
                className="form-control"
                value={template.get('_id')}
                options={templateOptions.toJS()}
                onChange={e => {
                  this.props.changeTemplate(this.props.model, e.target.value);
                }}
              />
            </li>
          </ul>
        </FormGroup>
      );
    }

    return (
      <ul className="search__filter">
        <div className="text-center protip">
          <Icon icon="lightbulb" /> <b>ProTip!</b>
          <span>
            You can create metadata templates in <I18NLink to="/settings">settings</I18NLink>.
          </span>
        </div>
      </ul>
    );
  }

  render() {
    const {
      model,
      template,
      templateOptions,
      id,
      multipleEdition,
      showSubset,
      version,
      storeKey,
      highlightedProps,
    } = this.props;

    if (!template) {
      return <div />;
    }

    const titleLabel = template.get('commonProperties')
      ? template
          .get('commonProperties')
          .find(p => p.get('name') === 'title')
          .get('label')
      : 'Title';

    return (
      <Form
        id={id}
        model={model}
        onSubmit={this.onSubmit}
        validators={validator.generate(template.toJS(), multipleEdition)}
        onSubmitFailed={this.onSubmitFailed}
      >
        {!multipleEdition && (!showSubset || showSubset.includes('title')) && (
          <FormGroup model=".title">
            <ul className="search__filter">
              <li>
                <label>
                  <Translate context={template.get('_id')}>{titleLabel}</Translate>{' '}
                  <span className="required">*</span>
                </label>
              </li>
              <li className="wide">
                <Field model=".title">
                  <textarea className="form-control" />
                </Field>
              </li>
              <IconField model={model} />
            </ul>
          </FormGroup>
        )}

        {(!showSubset || showSubset.includes('template')) &&
          this.renderTemplateSelect(templateOptions, template)}
        <MetadataFormFields
          multipleEdition={multipleEdition}
          thesauris={this.props.thesauris}
          model={model}
          template={template}
          showSubset={showSubset}
          version={version}
          highlightedProps={highlightedProps}
          storeKey={storeKey}
        />
      </Form>
    );
  }
}

MetadataForm.defaultProps = {
  id: 'metadataForm',
  multipleEdition: false,
  showSubset: undefined,
  version: undefined,
  initialTemplateId: undefined,
  componentWillUnmount: () => {},
  notify: () => {},
  changeTemplate: () => {},
  onSubmit: () => {},
  highlightedProps: [],
  storeKey: '',
};

MetadataForm.propTypes = {
  model: PropTypes.string.isRequired,
  template: PropTypes.instanceOf(Immutable.Map).isRequired,
  templateId: PropTypes.string,
  initialTemplateId: PropTypes.string,
  multipleEdition: PropTypes.bool,
  templateOptions: PropTypes.instanceOf(Immutable.List).isRequired,
  thesauris: PropTypes.instanceOf(Immutable.List).isRequired,
  changeTemplate: PropTypes.func,
  onSubmit: PropTypes.func,
  notify: PropTypes.func,
  id: PropTypes.string,
  showSubset: PropTypes.arrayOf(PropTypes.string),
  version: PropTypes.string,
  componentWillUnmount: PropTypes.func,
  highlightedProps: PropTypes.arrayOf(PropTypes.string),
  storeKey: PropTypes.string,
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ notify: notificationActions.notify }, dispatch);
}

export const mapStateToProps = (state, ownProps) => ({
  thesauris: ownProps.thesauris ? ownProps.thesauris : state.thesauris || Immutable.fromJS([]),
  template: ownProps.template
    ? ownProps.template
    : state.templates.find(tmpl => tmpl.get('_id') === ownProps.templateId) ||
      immutableDefaultTemplate,
  templateOptions: selectTemplateOptions(state),
});

export default connect(mapStateToProps, mapDispatchToProps)(MetadataForm);
