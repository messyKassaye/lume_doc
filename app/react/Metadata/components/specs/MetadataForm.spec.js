/** @format */

import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';
import { Form, Field } from 'react-redux-form';

import { MetadataForm, mapStateToProps } from '../MetadataForm';
import MetadataFormFields from '../MetadataFormFields';
import { Select as SimpleSelect } from '../../../Forms';

describe('MetadataForm', () => {
  let component;
  let fieldsTemplate;
  let props;
  let templates;

  beforeEach(() => {
    fieldsTemplate = [
      { name: 'field1', label: 'label1' },
      { name: 'field2', label: 'label2', type: 'select', content: '2' },
      { name: 'field3', label: 'label3', type: 'multiselect', content: '2' },
      { name: 'field4', label: 'label4', type: 'date' },
    ];

    templates = Immutable.fromJS([
      {
        name: 'template1',
        _id: 'templateId',
        properties: fieldsTemplate,
        commonProperties: [{ name: 'title', label: 'Title' }],
      },
      {
        name: 'template2',
        _id: '2',
        properties: [{ name: 'field3' }],
        commonProperties: [{ name: 'title', label: 'Title' }],
        isEntity: false,
      },
      {
        name: 'template3',
        _id: '3',
        properties: [{ name: 'field4' }],
        commonProperties: [{ name: 'title', label: 'Title' }],
        isEntity: true,
      },
    ]);

    props = {
      metadata: {
        _id: [{ value: 'docId' }],
        template: [{ value: 'templateId' }],
        title: [{ value: 'testTitle' }],
        metadata: [{ value: { field1: 'field1value', field2: 'field2value' } }],
      },
      templates,
      template: templates.get(0),
      templateOptions: Immutable.fromJS([{ label: 'template1', value: 'templateId' }]),
      thesauris: Immutable.fromJS([
        { _id: 2, name: 'thesauri', values: [{ label: 'option1', id: '1' }] },
      ]),
      onSubmit: jasmine.createSpy('onSubmit'),
      changeTemplate: jasmine.createSpy('changeTemplate'),
      model: 'metadata',
    };
  });

  const render = () => {
    component = shallow(<MetadataForm {...props} />);
  };

  describe('Icon field', () => {
    it('should remove icon', () => {});
  });

  it('should render a form with metadata as model', () => {
    render();
    const form = component.find(Form);
    expect(form.props().model).toEqual('metadata');
  });

  it('should display custom title label', () => {
    props.template = Immutable.fromJS({
      name: 'template4',
      _id: 'template4',
      commonProperties: [{ name: 'title', label: 'Name' }],
      properties: [],
    });
    render();
    const title = component
      .find('FormGroup')
      .first()
      .find({ children: 'Name' });
    expect(title.length).toBe(1);
  });

  it('should render MetadataFormFields passing thesauris state and template', () => {
    render();
    const formFields = component.find(MetadataFormFields);

    expect(formFields.props().thesauris).toBe(props.thesauris);
    expect(formFields.props().state).toBe(props.state);
    expect(formFields.props().template).toBe(props.templates.get(0));
  });

  it('should pass the model to Form and MetadataFormFields', () => {
    render();
    const form = component.find(Form);
    expect(form.props().model).toBe('metadata');
    const metadataFields = component.find(MetadataFormFields);
    expect(metadataFields.props().model).toBe('metadata');
  });

  it('should render title field as a textarea', () => {
    render();
    expect(component.find(Field)).toMatchSnapshot();
  });

  describe('on template change', () => {
    const warningMessage = 'Changing the type will erase all connections to this entity.';

    it('should call changeTemplate with the template', () => {
      render();
      const template = component.find(SimpleSelect).first();
      template.simulate('change', { target: { value: '2' } });
      expect(props.changeTemplate).toHaveBeenCalledWith(props.model, props.templates.toJS()[1]._id);
    });

    it('should warn about the loss of connections if the entity is already a saved one', () => {
      props.templateId = '2';
      props.initialTemplateId = 'templateId';
      render();
      const warning = component.find({ children: warningMessage });
      expect(warning.length).toBe(1);
    });

    it('should not warn about the loss of connections if the entity is a new one', () => {
      props.templateId = '2';
      props.initialTemplateId = undefined;
      render();
      const warning = component.find({ children: warningMessage });
      expect(warning.length).toBe(0);
    });

    it('should not warn about the loss of connections if the template is the same than the original one', () => {
      props.templateId = 'templateId';
      props.initialTemplateId = 'templateId';
      render();
      const warning = component.find({ children: warningMessage });
      expect(warning.length).toBe(0);
    });
  });

  describe('submit', () => {
    it('should call onSubmit with the values wrapped', () => {
      render();
      const UnwrapedEntity = {
        _id: '123',
        template: '456',
        language: 'en',
        metadata: {
          prop_one: ['one', 'two', 'three'],
          prop_two: 'Doctor who?',
          prop_three: null,
        },
        file: {},
        fullText: [],
      };
      const WrapedEntity = {
        _id: '123',
        template: '456',
        language: 'en',
        metadata: {
          prop_one: [{ value: 'one' }, { value: 'two' }, { value: 'three' }],
          prop_two: [{ value: 'Doctor who?' }],
          prop_three: [{ value: null }],
        },
      };
      component.find(Form).simulate('submit', UnwrapedEntity);
      expect(props.onSubmit).toHaveBeenCalledWith(WrapedEntity, 'metadata');
    });
  });

  describe('mapStateToProps', () => {
    let state;
    let ownProps;

    beforeEach(() => {
      state = { templates };
      ownProps = { templates, templateId: templates.get(1).get('_id') };
    });

    it('should return template based on metadata.template', () => {
      expect(mapStateToProps(state, ownProps).template).toBe(templates.get(1));
    });
  });
});
