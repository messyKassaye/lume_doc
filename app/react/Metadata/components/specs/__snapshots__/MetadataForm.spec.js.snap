// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataForm should render title field as a textarea 1`] = `
<Modeled(Connect(Field))
  model=".title"
>
  <textarea
    className="form-control"
  />
</Modeled(Connect(Field))>
`;
